### PR TITLE
8254260: Consider splitting binding recipe operators that serve a dual role

### PR DESCRIPTION
**Pull request by JornVernee**
Hi,

This patch splits the binding recipe operators that currently serve a dual purpose into 2 operators each:
- MOVE -> VM_STORE and VM_LOAD
- DEREFERENCE -> BUFFER_STORE and BUFFER_LOAD
- CONVERT_ADDRESS -> BOX_ADDRESS and UNBOX_ADDRESS

Note that I also added a TO_SEGMENT operator which converts a MemoryAddress into a MemorySegment. This was previously done as part of the COPY operator, but only for upcalls, so I had to split out this functionality so that COPY could have only one interpretation as well. This will also enable some more optimizations down the line, such as merely wrapping a by-value struct passed as a pointer in a MemorySegment, instead of having to make a copy.

This also regularized the use of allocators by the operators. Now an allocator is always passed as an argument, and that's what is used to do allocations, instead of relying implicitly on MemorySegment::allocateNative. (The index of the allocator argument is now also always passed when specializing an operator, to indicate the dependency, and so that operator impls don't have to 'guess' which argument index is the allocator (see for instance the implementation of Copy::specialize).

The rest of the patch is more or less a mechanical renaming, and updating the tests/CallArrangers to use the different operator names.

Thanks,
Jorn

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ❌ (2/5 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Build / test | ⏳ (1/1 running) |    |     | 
| Test (tier1) |    |  ⏳ (3/9 running) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x64 (build hotspot minimal)](https://github.com/JornVernee/panama-foreign/runs/1231457476)
- [Linux x64 (build hotspot zero)](https://github.com/JornVernee/panama-foreign/runs/1231457450)

### Issue
 * [JDK-8254260](https://bugs.openjdk.java.net/browse/JDK-8254260): Consider splitting binding recipe operators that serve a dual role

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/375/head:pull/375`
`$ git checkout pull/375`

### CODE REVIEW

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java [hunk 1]
```diff
@@ -281,22 +283,47 @@ private static void checkType(Class<?> type) {
...
         return copy;
     }
 
-    public static Move move(VMStorage storage, Class<?> type) {
+    private static MemorySegment allocateBuffer(long size, long allignment, SharedUtils.Allocator allocator) {
+        return allocator.allocate(size, allignment);
+    }
+
+    // FIXME should register with scope
+    private static MemorySegment toSegment(MemoryAddress operand, long size) {
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:303]: This seems to be only used inside the `ToSegment` binding, so, perhaps this could should be moved there (also because this is an interpreter function, not a binding factory - e.g. the overload with the `toSegment` binding factory is confusing)

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java [hunk 2]
```diff
@@ -324,19 +359,42 @@ public static Dup dup() {
         return new Binding.Builder();
     }
 
+    @Override
+    public boolean equals(Object o) {
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:363]: General question - where do we use equals/hash? Just for testing purposes right?

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java [hunk 3]
```diff
@@ -323,7 +340,9 @@ Object invokeMoves(Object[] args, Binding.Move[] argBindings, Binding.Move[] ret
     Object invokeInterpBindings(Object[] args, MethodHandle leaf,
                                 Map<VMStorage, Integer> argIndexMap,
                                 Map<VMStorage, Integer> retIndexMap) throws Throwable {
-        NativeScope scope = bufferCopySize != 0 ? NativeScope.boundedScope(bufferCopySize) : null;
+        boolean hasBufferCopies = bufferCopySize != 0;
+        NativeScope scope = hasBufferCopies ? NativeScope.boundedScope(bufferCopySize) : null;
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:344]: If Allocator had a close() method, this could be simplified a little?

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java [hunk 4]
```diff
@@ -200,58 +216,59 @@ private MethodHandle specialize(MethodHandle leafHandle) {
 
         MethodHandle specializedHandle = leafHandle; // initial
 
-        int insertPos = -1;
+        int argInsertPos = -1;
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:219]: The more we tweak this routine, the more I'm convinced we should just spin specialized bytecode :-) - it's becoming increasingly hard to follow what's happening here. (not for this review - more longer term)

File: test/jdk/java/foreign/NativeTestHelper.java [hunk 5]
```diff
@@ -33,7 +33,7 @@
     }
 
     public static boolean isIntegral(MemoryLayout layout) {
-        return kind(layout).isIntergral();
+        return kind(layout).isIntegral();
```
Comment (mcimadamore) [inline test/jdk/java/foreign/NativeTestHelper.java:36]: Whoops - sorry - I guess I fixed the API and have not run the tests?

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java [hunk 6]
```diff
@@ -218,21 +216,27 @@
...
+                    methodType(MemorySegment.class, MemoryAddress.class, long.class));
         } catch (ReflectiveOperationException e) {
             throw new RuntimeException(e);
         }
     }
 
     enum Tag {
-        MOVE,
-        DEREFERENCE,
+        VM_STORE,
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:230]: Should these be called REG_STORE and REG_LOAD? Ah, I guess we also have stack, right? Ok, I agree with the name - these are "moves" that only the VM can perform, makes sense.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java [hunk 7]
```diff
@@ -476,77 +562,111 @@ public long offset() {
...
+        }
+    }
+
+    /**
+     * BUFFER_STORE([offset into memory region], [type])
+     * Pops a MemorySegment from the operand stack, loads a [type] from
+     * [offset into memory region] from it, and pushes it onto the operand stack.
+     * The [type] must be one of byte, short, char, int, long, float, or double
+     */
+    public static class BuffferStore extends Dereference {
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:591]: Typo, there are 3 `f`

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java [hunk 8]
```diff
@@ -476,77 +562,111 @@ public long offset() {
...
+            MethodHandle setter = varHandle().toMethodHandle(VarHandle.AccessMode.SET);
+            setter = setter.asType(methodType(void.class, MemorySegment.class, type()));
+            return collectArguments(specializedHandle, insertPos + 1, setter);
         }
 
-        private VarHandle varHandle() {
-            return MemoryHandles.insertCoordinates(MemoryHandles.varHandle(type, ByteOrder.nativeOrder()), 1, offset);
+        @Override
+        public String toString() {
+            return "BuffferStore{" +
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:622]: Typo here too

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java [hunk 9]
```diff
@@ -90,7 +90,12 @@ private static void verifyUnboxBindings(Class<?> inType, List<Binding> bindings)
         stack.push(inType);
 
         for (Binding b : bindings) {
-            b.verifyUnbox(stack);
+            if (b.tag() == Binding.Tag.VM_LOAD
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java:93]: We could have static EnumSet for what's an box vs. unbox binding, so that you can filter by checking for presence in the right enum set?

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java [hunk 10]
```diff
@@ -102,7 +107,11 @@ private static void verifyBoxBindings(Class<?> expectedReturnType, List<Binding>
         Deque<Class<?>> stack = new ArrayDeque<>();
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java:107]: I've seen in the code that the verifyUnbox method speaks about a `returnType` parameter - should it be an `outType` for symmetry (since, I guess, we use this more than just for returns - e.g. we use it for args in upcalls).

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java [hunk 11]
```diff
@@ -98,7 +99,7 @@ private void invoke(MemoryAddress buffer) {
                                 ? stackArgsBase.asSlice(storage.index() * abi.arch.typeSize(abi.arch.stackType()))
                                 : bufferBase.asSlice(layout.argOffset(storage));
                             return SharedUtils.read(ptr, type);
-                        });
+                        }, DEFAULT_ALLOCATOR);
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java:102]: For upcalls, couldn't we use a `NativeScope` and close at the end of the call? If the callback leaks segments outside... well, too bad.

**Review by mcimadamore — APPROVED** [review]
Looks a great cleanup. Left some comments.
I also ran all tests here and everything looks good.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java (hunk 1 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:303]: Yeah, good idea. This and 2 others are used for the `specialize` implemantion, but they can be moved inside the binding classes.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java (hunk 2 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:363]: Yeah, we use them at least equals for the CallArranger tests, which is comparing arrays of Bindings.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java (hunk 3 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:344]: This is not really a problem with Allocator, it's a problem with having either a NativeScope, or having none (null), which will result in an NPE when trying to reference `scope::allocateNative`. That's why this code is conditional.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java (hunk 4 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:219]: Yeah, especially the indexing is tricky since we have to walk the recipe in reverse, but we shouldn't have that problem when spinning bytecode.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java (hunk 6 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:230]: REG_LOAD and REG_STORE also works for me. The name is derived from VMReg in the HotSpot code -> VMStorage in our code -> VM_LOAD/VM_STORE. (So, REG works equally well).

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java (hunk 7 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:591]: Thanks!

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java (hunk 9 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java:93]: Yeah, that's a good idea.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java (hunk 11 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java:102]: Yes, exactly. But, seemed like a separate patch.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java (hunk 10 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/CallingSequenceBuilder.java:107]: I'll change the name. FWIW, `returnType` refers to the type 'returned' by the chain of operations.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java (hunk 3 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:344]: I'm experimenting with this a bit. Seems like there is some room for improvement.

**Comment by JornVernee** [comment]
@mcimadamore Thanks for the review!

I've added the suggested close method to SharedUtils.Allocator (see: https://github.com/openjdk/panama-foreign/pull/375/commits/9ffebc0016672e273a066c84f8cafdae55bb10c4). Looks like an improvement?

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java [hunk 12]
```diff
@@ -215,11 +215,11 @@
                     methodType(MemoryAddress.class, long.class));
             MH_BASE_ADDRESS = lookup.findVirtual(MemorySegment.class, "address",
                     methodType(MemoryAddress.class));
-            MH_COPY_BUFFER = lookup.findStatic(Binding.class, "copyBuffer",
+            MH_COPY_BUFFER = lookup.findStatic(Binding.Copy.class, "copyBuffer",
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:218]: I'm assuming these are only really required in the specialize() method of the various subclasses - should we move those constants there too?

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java [hunk 13]
```diff
@@ -340,10 +341,10 @@ Object invokeMoves(Object[] args, Binding.VMStore[] argBindings, Binding.VMLoad[
     Object invokeInterpBindings(Object[] args, MethodHandle leaf,
                                 Map<VMStorage, Integer> argIndexMap,
                                 Map<VMStorage, Integer> retIndexMap) throws Throwable {
-        boolean hasBufferCopies = bufferCopySize != 0;
-        NativeScope scope = hasBufferCopies ? NativeScope.boundedScope(bufferCopySize) : null;
-        SharedUtils.Allocator unboxAllocator = hasBufferCopies ? SharedUtils.Allocator.ofScope(scope) : null;
-        try {
+        SharedUtils.Allocator unboxAllocator = bufferCopySize != 0
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:344]: Nice! I also like THROWING_ALLOCATOR more than `null`

**Review by mcimadamore — APPROVED** [review]
Very nice - just added another minor stylistic comment on binding MH constants

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java (hunk 12 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:218]: We can, though we end up duplicating the try/catch & lookup code if we do, or would have to add some helper method, for both findStatic and findVirtual. That's why I held off on doing that.
